<a href="https://colab.research.google.com/github/AshutoshAgrahari/GoogleColab/blob/master/Intro_to_Computer_Vision.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Introduction to Computer Vision
### Ref: https://github.com/zalandoresearch/fashion-mnist
https://colab.research.google.com/github/lmoroney/dlaicourse/blob/master/Course%201%20-%20Part%204%20-%20Lesson%202%20-%20Notebook.ipynb

In [0]:
# Loading the tensorflow and keras
import tensorflow as tf
from tensorflow import keras

In [0]:
#prevent numpy exponential 
#notation on print, default False
import numpy as np
np.set_printoptions(suppress = True)

Calling load_data on this object will give you two sets of two lists, these will be the training and testing values for the graphics that contain the clothing items and their labels.

In [0]:
# Loading Fashion-MNIST dataset from https://github.com/zalandoresearch/fashion-mnist
fashion_mnist = keras.datasets.fashion_mnist
(train_images, train_labels),(test_images,test_labels) = fashion_mnist.load_data()

4423680/4422102 [==============================] - 0s 0us/step


What does these values look like? Let's print a training image, and a training label to see...Experiment with different indices in the array. For example, also take a look at index 42...that's a a different boot than the one at index 0

In [0]:
import matplotlib.pyplot as plt

In [0]:
plt.imshow(train_images[0])
#print(train_labels[0])
#print(train_images[0])

You'll notice that all of the values in the number are between 0 and 255. If we are training a neural network, for various reasons it's easier if we treat all values as between 0 and 1, a process called 'normalizing'...and fortunately in Python it's easy to normalize a list like this without looping. You do it like this:

In [0]:
train_images = train_images/255
test_images = test_images/255

Let's now design the model.

In [0]:
model = tf.keras.models.Sequential([tf.keras.layers.Flatten(),
                                   tf.keras.layers.Dense(128, activation=tf.nn.relu),
                                   tf.keras.layers.Dense(10, activation=tf.nn.softmax)])

**Sequential:** that defines a SEQUENCE of layers in neural network.

**Flatten:** Images data has been stored in 2-dimensional matrix which needs to convert into 1-dimensional vector by flatten the square set into vector by rows.

**Dense:** add a layer of neurons

Each layer of neurons need an **activation function** to tell them what to do. It has many options available to choose an activation function.

**Relu:** effectively mean "if X >0 then return X, else return 0" -- so what it does, it only passes values 0 or greater than to next layer in the network.

**Softmax:** takes a set of values and effectively pick the biggest one,  so for example, if the output of the last layer looks like  [0.1, 0.1, 0.05, 0.1, 9.5, 0.1, 0.05, 0.05, 0.05], it saves from fishing through it looking for the biggest value, and turns it into [0,0,0,0,1,0,0,0,0] -- The goal is to save a lot of coding!

The next thing to do, now the model is defined, is to actually build it. You do this by compiling it with an optimizer and loss function as before -- and then you train it by calling model.fit asking it to fit your training data to your training labels -- i.e. have it figure out the relationship between the training data and its actual labels, so in future if you have data that looks like the training data, then it can make a prediction for what that data would look like.

In [0]:
model.compile(optimizer= tf.train.AdamOptimizer(),
             loss = 'sparse_categorical_crossentropy',
             metrics = ['accuracy'])

Instructions for updating:
Colocations handled automatically by placer.


In [0]:
model.fit(train_images, train_labels, epochs=15 )

Epoch 1/15
60000/60000 [==============================] - 5s 77us/sample - loss: 0.2817 - acc: 0.8955
Epoch 2/15
60000/60000 [==============================] - 5s 76us/sample - loss: 0.2694 - acc: 0.9003
Epoch 3/15
60000/60000 [==============================] - 5s 76us/sample - loss: 0.2584 - acc: 0.9044
Epoch 4/15
60000/60000 [==============================] - 5s 77us/sample - loss: 0.2501 - acc: 0.9068
Epoch 5/15
60000/60000 [==============================] - 5s 79us/sample - loss: 0.2405 - acc: 0.9104
Epoch 6/15
60000/60000 [==============================] - 5s 77us/sample - loss: 0.2326 - acc: 0.9131
Epoch 7/15
60000/60000 [==============================] - 5s 77us/sample - loss: 0.2254 - acc: 0.9153
Epoch 8/15
60000/60000 [==============================] - 5s 77us/sample - loss: 0.2166 - acc: 0.9186
Epoch 9/15
60000/60000 [==============================] - 5s 77us/sample - loss: 0.2118 - acc: 0.9221
Epoch 10/15
60000/60000 [==============================] - 5s 77us/sample - loss: 

In [0]:
model.evaluate(test_images, test_labels)

10000/10000 [==============================] - 0s 43us/sample - loss: 0.3717 - acc: 0.8787


[0.37169571934938433, 0.8787]

For me, that returned a accuracy of about .8787, which means it was about 87% accurate. As expected it probably would not do as well with *unseen* data as it did with data it was trained on! 

#Exploration Exercises

###Exercise 1:
For this first exercise run the below code: It creates a set of classifications for each of the test images, and then prints the first entry in the classifications. The output, after you run it is a list of numbers. Why do you think this is, and what do those numbers represent? 

In [0]:
classifications = model.predict(test_images)

In [0]:
print(classifications[0:2])
#classifications[0]

[[0.         0.         0.         0.         0.         0.00000212
  0.         0.00147068 0.         0.99852717]
 [0.00004924 0.         0.9997949  0.         0.00014319 0.00000002
  0.00001267 0.         0.         0.        ]]


In [0]:
print(test_labels[0:2])

[9 2]


model.predict will return the array with 10 numbers which is equal to labels. We have 10 different fashion garments images to recognise by machine.

We can get the classified label from classifications as position of greatest number. For classification[0], we have hightest number is 0.99852717 at 9 position(counted from 0)

##Exercise 2: 
Let's now look at the layers in your model. Experiment with different values for the dense layer with 512 neurons. What different results do you get for loss, training time etc? Why do you think that's the case? 



In [93]:
import tensorflow as tf
print(tf.__version__)

class myCallback(tf.keras.callbacks.Callback):
  def on_epoch_end(self, epoch, logs={}):
    if(logs.get('loss')< 14):
      print("\nReached 60% accuracy so cancelling training!")
      self.model.stop_training = True

callbacks = myCallback()

Fashion_mnist = tf.keras.datasets.fashion_mnist
#mnist = tf.keras.datasets.mnist
(training_images, training_labels), (test_images, test_labels) = Fashion_mnist.load_data()

training_image = training_images/255.0
test_images = test_images/255.0

#print(test_images[0])

model = tf.keras.models.Sequential([tf.keras.layers.Flatten(input_shape=(28, 28)),
                                   #tf.keras.layers.Dense(128, activation = tf.nn.relu),
                                   tf.keras.layers.Dense(512, activation = tf.nn.relu), 
                                   tf.keras.layers.Dense(10, activation = tf.nn.softmax)])

model.compile(#optimizer= 'adam',
              optimizer=tf.train.AdamOptimizer(),
             loss = 'sparse_categorical_crossentropy'
 #            metrics = ['accuracy']
)

model.fit(training_images, training_labels, epochs= 5, callbacks = [callbacks])

#model.evaluate(test_images, test_labels)

#classifications = model.predict(test_images)

#print(classifications[1000])
#print(test_labels[1000])

1.13.1
Epoch 1/5
59648/60000 [============================>.] - ETA: 0s - loss: 13.2475
Reached 60% accuracy so cancelling training!
60000/60000 [==============================] - 12s 198us/sample - loss: 13.2437


In [84]:
import tensorflow as tf

class myCallback(tf.keras.callbacks.Callback):
  def on_epoch_end(self, epoch, logs={}):
    if(logs.get('acc')>0.85):
      print("\nReached 60% accuracy so cancelling training!")
      self.model.stop_training = True

mnist = tf.keras.datasets.fashion_mnist

(x_train, y_train),(x_test, y_test) = mnist.load_data()
x_train, x_test = x_train / 255.0, x_test / 255.0

callbacks = myCallback()

model = tf.keras.models.Sequential([
  tf.keras.layers.Flatten(input_shape=(28, 28)),
  tf.keras.layers.Dense(256, activation=tf.nn.relu),
  tf.keras.layers.Dense(10, activation=tf.nn.softmax)
])
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

model.fit(x_train, y_train, epochs=10, callbacks=[callbacks])

Epoch 1/10
60000/60000 [==============================] - 10s 162us/sample - loss: 0.4819 - acc: 0.8306
Epoch 2/10
59904/60000 [============================>.] - ETA: 0s - loss: 0.3644 - acc: 0.8672
Reached 60% accuracy so cancelling training!
60000/60000 [==============================] - 12s 197us/sample - loss: 0.3643 - acc: 0.8672
